node/src/components/gossiper/gossip_table.rs

```diff
@@ -704,6 +716,54 @@ mod tests {
         assert!(gossip_table.finished.contains(&data_id));
     }
 
+    #[test]
+    fn should_terminate_via_incoming_gossip() {
+        let mut rng = crate::new_rng();
+        let node_ids = random_node_ids(&mut rng);
+        let data_id1: u64 = rng.gen();
+        let data_id2: u64 = rng.gen();
+
+        let mut gossip_table = GossipTable::new(Config::default());
+
+        let limit = EXPECTED_DEFAULT_HOLDERS_LIMIT - 1;
+        for node_id in node_ids.iter().take(limit) {
+            let _ = gossip_table.new_partial_data(&data_id1, node_id.clone());
+            assert!(!gossip_table.finished.contains(&data_id1));
+
+            let _ = gossip_table.new_complete_data(&data_id2, Some(node_id.clone()));
+            assert!(!gossip_table.finished.contains(&data_id2));
+        }
+
+        let _ = gossip_table
+            .new_partial_data(&data_id1, node_ids[EXPECTED_DEFAULT_HOLDERS_LIMIT].clone());
+        assert!(gossip_table.finished.contains(&data_id1));
+
+        let _ = gossip_table.new_complete_data(
+            &data_id2,
+            Some(node_ids[EXPECTED_DEFAULT_HOLDERS_LIMIT].clone()),
+        );
+        assert!(gossip_table.finished.contains(&data_id2));
+    }
+
+    #[test]
+    fn should_terminate_via_checking_timeout() {
+        let mut rng = crate::new_rng();
+        let node_ids = random_node_ids(&mut rng);
+        let data_id: u64 = rng.gen();
+
+        let mut gossip_table = GossipTable::new(Config::default());
+
+        let limit = EXPECTED_DEFAULT_HOLDERS_LIMIT - 1;
+        for node_id in node_ids.iter().take(limit) {
+            let _ = gossip_table.new_complete_data(&data_id, Some(node_id.clone()));
+            assert!(!gossip_table.finished.contains(&data_id));
+        }
+
+        let _ =
+            gossip_table.check_timeout(&data_id, node_ids[EXPECTED_DEFAULT_HOLDERS_LIMIT].clone());
+        assert!(gossip_table.finished.contains(&data_id));
+    }
```

Comment:
I don't understand why this unit test is supposed to pass. It says that it expects entries to be moved to `finished` b/c of the timeout but I don't see the timeout being set here anywhere. I don't see that we're triggering the timeout to happen etc.

Reply:
Yeah, the gossip table is controlled entirely by the gossiper component.  The gossiper deals with setting timeouts and asking the network to send messages, etc.

The gossiper would call `gossip_table.check_timeout` when its timeout triggered, so that's why there's no actual timeouts in the gossip table tests.